Non-Annotator DecCBOR instances

teodanciu opened the pull request:
# Description

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

Add `DecCBOR` instances for types that have `DecCBOR Annotator` instances,  by leveraging the `Maybe ByteString`  in versioned Decoder. 

Closes: https://github.com/IntersectMBO/cardano-ledger/issues/4818


```
binary 
	PV1.Data                     EXISTS
core 
	BootstrapWitness             DONE (with new raw type and decodeMemoized)
	WitVKey kr                   DONE (with new raw type and decodeMemoized)
	MemoBytes t era              EXISTS
	PlutusData era               DONE (with fromPlainDecoder)
	Data era                     DONE (deriving newtype)
	PlutusBinary                 EXISTS
	Block h era                  DONE (with new raw type and decodeMemoized)
tpraos 
	BHeader c                    DONE
shelley 
	ShelleyTxRaw era             DONE
	ShelleyTx era                DONE
	ShelleyTxSeq era             DONE 
	MultiSig era                 DONE
	MultiSigRaw era              DONE
	ShelleyTxAuxDataRaw era      EXISTS
	ShelleyTxAuxData era         DONE (deriving newtype)
	ShelleyTxBodyRaw era         EXISTS
	ShelleyTxBody era            DONE (deriving newtype) 
	ShelleyTxWitsRaw era         DONE
	ShelleyTxWits era            DONE
	LaxBlock h era               DONE 
allegra 
	AllegraTxBodyRaw () era      EXISTS
	AllegraTxBody era            DONE (deriving newtype) 
	TimelockRaw era              DONE
	Timelock era                 DONE (with decodeMemoized)
	AllegraTxAuxDataRaw era      DONE
	AllegraTxAuxData era         DONE (deriving newtype)
mary 
	MaryTxBodyRaw era            EXISTS
	MaryTxBody era               DONE (deriving newtype) 
alonzo 
	AlonzoTxBody era             DONE (deriving newtype) 
	AlonzoTxBodyRaw era          EXISTS
	AlonzoTxSeq era              DONE        
	AlonzoScript era             DONE
	AlonzoTx era                 DONE  
	AlonzoTxAuxDataRaw era       DONE 
	AlonzoTxAuxData era          DONE 
	TxDatsRaw era                DONE
	TxDats era                   DONE 
	RedeemersRaw era             DONE
	Redeemers era                DONE
	AlonzoTxWitsRaw era          DONE
	AlonzoTxWits era             DONE
	TranslationInstance era      DONE
babbage
	BabbageTxBody era            DONE 
	BabbageTxBodyRaw era         EXISTS
conway
	ConwayTxBodyRaw era          EXISTS 
	ConwayTxBody era             DONE
```


# Checklist

- [ ] Commits in meaningful sequence and with useful messages
- [ ] Tests added or updated when needed
- [ ] `CHANGELOG.md` files updated for packages with externally visible changes<br>
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] Version bounds in `.cabal` files updated when necessary<br>
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code formatted (use `scripts/fourmolize.sh`)
- [x] Cabal files formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

teodanciu committed:
Add `DecCBOR` instances for types in `cardano-ledger-core`

`PlutusData`, `Data`, `BootstrapWitness`, `WitVKey`, `Block`

teodanciu committed:
Add `DecCBOR` instances for types in `cardano-ledger-allegra`

`Timelock`, `AllegraTxAuxData`, `AllegraTxBody`

teodanciu committed:
Add `DecCBOR` instances for types in `cardano-ledger-shelley`

`ShelleyTxWits`, `ShelleyTxAuxData`, `ShelleyTxBody`, `ShelleyTx`, `ShelleyTxSeq`, `LaxBlock`

teodanciu committed:
Add `DecCBOR` instance for `BHeader`

teodanciu committed:
Add `DecCBOR` instances for types in `cardano-ledger-alonzo`

teodanciu committed:
Extract local bindings used to build TxWits decoder as function

so that they can be reused

teodanciu committed:
Remove redundant constraint from DecCBOR instance for AlonzoTxWitsRaw

teodanciu committed:
Add `DecCBOR` instance for `BabbageTxBody`

teodanciu committed:
Add `DecCBOR` instance for `ConwayTxBody`

teodanciu committed:
Rename `segwitTx` to distinguish `Annotator` from simple version

lehins commented on .github/workflows/haskell.yml:
```diff
@@ -213,7 +213,7 @@ jobs:
         - small-steps
         - vector-map
-        ghc: ["8.10.7", "9.2.8", "9.6.6", "9.8.4", "9.10.1"]
+        ghc: ["9.2.8", "9.6.6", "9.8.4", "9.10.1"]
```
We can't drop ghc-8.10.7 yet. See #4798

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -449,6 +449,23 @@ instance
+        <! From
+        <! From
+        <! From
+        <! D (maybeToStrictMaybe <$> decodeNullMaybe decCBOR)
```
```suggestion
        <! D (decodeNullStrictMaybe decCBOR)
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/BlockChain.hs:
```diff
@@ -233,12 +238,45 @@ txSeqDecoder lax = do
+  decCBOR = do
+    Annotated bodies bodiesBs <- decodeAnnotated decCBOR
+    Annotated wits witsBs <- decodeAnnotated decCBOR
+    Annotated auxDataMap auxDataBs <- decodeAnnotated decCBOR
```
I suggest a better naming convention. IMHO Bytes is better than some BS :smile: 
```suggestion
    Annotated bodies bodiesBytes <- decodeAnnotated decCBOR
    Annotated wits witsBytes <- decodeAnnotated decCBOR
    Annotated auxDataMap auxDataBytes <- decodeAnnotated decCBOR
```

lehins reviewed (commented):
Do no look at comments :smile:

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Tx/Internal.hs:
```diff
@@ -376,6 +399,33 @@ segwitTx
+segWitTx
+  body'
+  witnessSet
+  metadata =
```
```suggestion
segWitTx body' witnessSet metadata =
```

teodanciu committed:
Add equivalent non-Annotator cbor roundtrip tests

for all the ones using `DecCBOR (Annotator t)` instances

teodanciu committed:
Add check for non-Annotator version of golden serialsation tests

teodanciu committed:
Add functions to test the equivalence of `DecCBOR` instances

between the Annotator and simple version, both via arbitrary data
generation and via cddl and huddle

teodanciu committed:
Move Arbitrary instance for `PV1.Data` to core

teodanciu committed:
Move Arbitrary instance for `Data`, `BinaryData` and `Datum` to core

teodanciu committed:
Add tests to check equivalence of `DecCBOR` instances via `Arbitrary` data

teodanciu committed:
Add tests to check equivalence of `DecCBOR`instances via cddl and Huddle

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Keys/WitVKey.hs:
```diff
@@ -97,6 +100,24 @@ instance Typeable kr => DecCBOR (Annotator (WitVKey kr)) where
+instance Typeable kr => DecCBOR (WitVKey kr) where
+  decCBOR = do
+    mb <- decodeMemoized (decCBOR @(WitVKeyRaw kr))
+    let (Memo (WitVKeyRaw k s kh) bs) = mb
```
```suggestion
    Memo (WitVKeyRaw k s kh) bs <- decodeMemoized (decCBOR @(WitVKeyRaw kr))
```

lehins commented on libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary.hs:
```diff
@@ -0,0 +1,78 @@
+    -- because some arbitrary instances generate data that fails serialisation for some protocols
+    -- (for example, TxDats in Conway)
+    (Right _, Right _) -> decAnn `shouldBe` dec
+    _ -> pure ()
```
That is not good enough, because it can be that one of them succeeds, while another fails, which would be really bad.
```suggestion
  case (decAnn, dec) of
    -- we only check in case of successful deserialisation,
    -- because some arbitrary instances generate data that fails serialisation for some protocols
    -- (for example, TxDats in Conway)
    (Right _, Right _) -> decAnn `shouldBe` dec
    (Left _, Left _) -> pure ()
    _ -> error "MISMATCH! ......"
```

lehins commented on libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary.hs:
```diff
@@ -0,0 +1,78 @@
+  property $
+    forM_ [fromVersion .. toVersion] $ \version -> do
+      let bs = serialize version t
+      decoderEquivalenceExpectation @t version bs
```
You don't really need `decoderEquivalenceExpectation`, this should do the trick. In fact it will be betters, since it will test more stuff and produce more informative failures

```suggestion
      embedTripAnnExpectation version version shouldBe t
```

teodanciu replied:
That's great! Also allows me remove a bunch of constraints in other functions. 
I still have to leave the definition of `decoderEquivalenceExpectation` because it's used in the cddl specs .

teodanciu replied:
Interestingly, with this change I have some test failures for ` TxDats ConwayEra` , `Redeemers ConwayEra ` and for the Block tests. I will investigate why

lehins commented on libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/Arbitrary.hs:
```diff
@@ -208,6 +210,22 @@ deriving instance Arbitrary BlockNo
+                  <*> listOf (gendata (n `div` 2))
+              , PV1.List <$> listOf (gendata (n `div` 2))
+              ]
+      gendata _ = oneof [PV1.I <$> arbitrary, PV1.B <$> arbitrary]
```
```suggestion
        | otherwise = oneof [PV1.I <$> arbitrary, PV1.B <$> arbitrary]
```

lehins commented on libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/Arbitrary.hs:
```diff
@@ -208,6 +210,22 @@ deriving instance Arbitrary BlockNo
+instance Arbitrary PV1.Data where
+  arbitrary = resize 5 (sized gendata)
+    where
+      gendata n
```
Since you moved this `Arbitrary` instance let's fix it up a bit. For the sake of convention:
```suggestion
      genData n
```
As I mentioned in another comment, let's move it to `cardano-ledger-core` instead

lehins commented on libs/cardano-ledger-binary/cardano-ledger-binary.cabal:
```diff
@@ -129,6 +130,7 @@ library testlib
     hspec,
     hspec-core,
     iproute,
+    plutus-ledger-api,
```
It would be better to move testing of plutus related stuff into `cardano-ledger-core` instead, because Byron era packages also depend on `cardano-ledger-binary` and there is no need for Byron era to have a transitive dependency on the whole of plutus.

teodanciu replied:
Makes sense! I chose binary because the DecCBOR instance is defined there, but your reason is solid. Done, thanks

lehins commented on libs/cardano-ledger-binary/test/Test/Cardano/Ledger/Binary/BinarySpec.hs:
```diff
@@ -0,0 +1,13 @@
+spec :: Spec
+spec = do
+  describe "DecCBOR instances equivalence" $ do
+    decoderEquivalenceSpec @PV1.Data minBound maxBound
```
As mentioned in another comment, let's move this test into `cardano-ledger-core` instead

lehins commented on eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs:
```diff
@@ -204,6 +204,17 @@ instance Era era => DecCBOR (Annotator (TimelockRaw era)) where
+      decRaw 3 = SumD MOfN <! From <! D decCBOR
+      decRaw 4 = SumD TimeStart <! From
+      decRaw 5 = SumD TimeExpire <! From
+      decRaw n = Invalid n
```
Feel free to ignore this suggestion. It just I like this approach better since it contains less noise
```suggestion
  decCBOR = decode $ Summands "Timelock" $ \case
      0 -> SumD Signature <! From
      1 -> SumD AllOf <! D decCBOR
      2 -> SumD AnyOf <! D decCBOR
      3 -> SumD MOfN <! From <! D decCBOR
      4 -> SumD TimeStart <! From
      5 -> SumD TimeExpire <! From
      n -> Invalid n
```

lehins commented on eras/allegra/impl/src/Cardano/Ledger/Allegra/TxAuxData.hs:
```diff
@@ -187,6 +187,30 @@ instance Era era => DecCBOR (Annotator (AllegraTxAuxDataRaw era)) where
+      TypeListLen -> decodeFromList
+      TypeListLen64 -> decodeFromList
+      TypeListLenIndef -> decodeFromList
+      _ -> error "Failed to decode AuxiliaryData"
```
Yikes! Could you please fix this for the Annotator instance too please.
```suggestion
      _ -> fail "Failed to decode AllegraTxAuxDataRaw"
```

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxAuxData.hs:
```diff
@@ -232,6 +220,56 @@ instance Era era => DecCBOR (Annotator (AlonzoTxAuxDataRaw era)) where
+    TypeMapLenIndef -> decodeShelley
+    TypeListLen -> decodeShelleyMA
+    TypeListLen64 -> decodeShelleyMA
+    TypeListLenIndef -> decodeShelleyMA
```
ShelleyMA is a old named for MarryAllegra name. Let's try to avoid it
```suggestion
    TypeListLen -> decodeAllegra
    TypeListLen64 -> decodeAllegra
    TypeListLenIndef -> decodeAllegra
```

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxAuxData.hs:
```diff
@@ -232,6 +220,56 @@ instance Era era => DecCBOR (Annotator (AlonzoTxAuxDataRaw era)) where
+      decodeShelley =
+        decode
+          (Emit AlonzoTxAuxDataRaw <! From <! Emit StrictSeq.empty <! Emit Map.empty)
+      decodeShelleyMA =
```
Ditto
```suggestion
      decodeAllegra =
```

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxSeq/Internal.hs:
```diff
@@ -230,6 +232,65 @@ instance AlonzoEraTx era => DecCBOR (Annotator (AlonzoTxSeq era)) where
+        inRange x = (0 <= x) && (x <= (b - 1))
+        w = length wits
+    unless
+      (all inRange (Map.keysSet auxDataMap))
```
It is redundant to make a whole new `Set` to just conjoin on the keys. This will be more efficient from memory and time perspective (you 'll need to import `Data.Monoid (All(..))`): 
```suggestion
      (getAll (Map.foldMapWithKey (\k _ -> All (inRange k)) auxDataMap))
```

teodanciu replied:
TIL, thanks! Will change it for the `Annotator (DecCBOR)` to

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxSeq/Internal.hs:
```diff
@@ -230,6 +232,65 @@ instance AlonzoEraTx era => DecCBOR (Annotator (AlonzoTxSeq era)) where
+          )
+      )
+    let auxData = maybeToStrictMaybe <$> indexLookupSeq b auxDataMap
+    Annotated isValidIdxs isValidBs <- decodeAnnotated decCBOR
```
```suggestion
    Annotated isValidIdxs isValidBytes <- decodeAnnotated decCBOR
```

teodanciu replied:
I removed many but sorry, I missed some of the bs!  Done

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxSeq/Internal.hs:
```diff
@@ -230,6 +232,65 @@ instance AlonzoEraTx era => DecCBOR (Annotator (AlonzoTxSeq era)) where
+    Annotated bodies bodiesBytes <- decodeAnnotated decCBOR
+    Annotated wits witsBytes <- decodeAnnotated decCBOR
+    Annotated auxDataMap auxDataBytes <- decodeAnnotated decCBOR
+    let b = length bodies
```
`b`s and `w`s is the naming of the past :wink: 
```suggestion
    let bodiesLength = length bodies
```

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxSeq/Internal.hs:
```diff
@@ -230,6 +232,65 @@ instance AlonzoEraTx era => DecCBOR (Annotator (AlonzoTxSeq era)) where
+    Annotated auxDataMap auxDataBytes <- decodeAnnotated decCBOR
+    let b = length bodies
+        inRange x = (0 <= x) && (x <= (b - 1))
+        w = length wits
```
```suggestion
        witsLength = length wits
```

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxSeq/Internal.hs:
```diff
@@ -230,6 +232,65 @@ instance AlonzoEraTx era => DecCBOR (Annotator (AlonzoTxSeq era)) where
+          mkBasicTx body
+            & witsTxL .~ wt
+            & auxDataTxL .~ ad
+            & isValidTxL .~ isValid
```
```suggestion
    let mkTx body wits isValid auxData =
          mkBasicTx body
            & witsTxL .~ wits
            & auxDataTxL .~ auxData
            & isValidTxL .~ isValid
```

lehins reviewed (commented):
I've started reviewing this PR today and it looks great so far.
I'll finish reviewing tomorrow, but I wanted to submit the feedback that I have so far.

teodanciu committed:
Fix handling of deserialization failure in `AllegraTxAuxDataRaw`

teodanciu committed:
Small improvements and better names in `Annotator AlonzoTxSeq` decoder

teodanciu commented on libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Encoding/EncCBOR.hs:
```diff
@@ -1062,8 +1063,10 @@ deriving instance EncCBOR EpochInterval
 
-instance EncCBOR PV1.Data where
-  encCBOR = fromPlainEncoding . Serialise.encode
+instance Plain.ToCBOR PV1.Data where
```
This is necessary in order to use it with `embedTripAnnExpectation`. Is there a better place to place it?

lehins replied:
Yes, `Cardano.Ledger.Plutus.Data`.

I guess we already depend on plutus in `cardano-ledger-binary`.
It is up to you if you wanna leave it all here or move all of those plutus instances into `cardano-ledger-core`. Either way is fine.

teodanciu committed:
Add `DecCBOR` instance for `MaryTxBody`

lehins commented on eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs:
```diff
@@ -204,6 +204,16 @@ instance Era era => DecCBOR (Annotator (TimelockRaw era)) where
+    0 -> SumD Signature <! From
+    1 -> SumD AllOf <! D decCBOR
+    2 -> SumD AnyOf <! D decCBOR
+    3 -> SumD MOfN <! From <! D decCBOR
```
Simplification
```suggestion
    1 -> SumD AllOf <! From
    2 -> SumD AnyOf <! From
    3 -> SumD MOfN <! From <! From
```

lehins commented on eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs:
```diff
@@ -204,6 +204,16 @@ instance Era era => DecCBOR (Annotator (TimelockRaw era)) where
       decRaw n = Invalid n
 
+instance Era era => DecCBOR (TimelockRaw era) where
+  decCBOR = decode $ Summands "Timelock" $ \case
```
```suggestion
  decCBOR = decode $ Summands "TimelockRaw" $ \case
```

lehins commented on eras/allegra/impl/src/Cardano/Ledger/Allegra/TxAuxData.hs:
```diff
@@ -187,6 +187,30 @@ instance Era era => DecCBOR (Annotator (AllegraTxAuxDataRaw era)) where
+        decode
+          ( RecD AllegraTxAuxDataRaw
+              <! From
+              <! D decCBOR
```
```suggestion
              <! From
```

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxAuxData.hs:
```diff
@@ -232,6 +220,56 @@ instance Era era => DecCBOR (Annotator (AlonzoTxAuxDataRaw era)) where
+          (Emit AlonzoTxAuxDataRaw <! From <! Emit StrictSeq.empty <! Emit Map.empty)
+      decodeAllegra =
+        decode
+          (RecD AlonzoTxAuxDataRaw <! From <! D (decodeStrictSeq decCBOR) <! Emit Map.empty)
```
```suggestion
          (RecD AlonzoTxAuxDataRaw <! From <! From <! Emit Map.empty)
```

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxAuxData.hs:
```diff
@@ -232,6 +220,56 @@ instance Era era => DecCBOR (Annotator (AlonzoTxAuxDataRaw era)) where
+      auxDataField 1 =
+        field
+          (\x ad -> ad {atadrTimelock = atadrTimelock ad <> x})
+          (D (decodeStrictSeq decCBOR))
```
```suggestion
      auxDataField 1 = field (\x ad -> ad {atadrTimelock = atadrTimelock ad <> x}) From
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxBody/Internal.hs:
```diff
@@ -267,6 +267,14 @@ instance
+deriving newtype instance
+  ( EraPParams era
+  , DecCBOR (TxOut era)
+  , ShelleyEraTxCert era
+  , TxCert era ~ ConwayTxCert era
```
It should be this constraint instead. It was defined using a wrong type. See #4899 
```suggestion
  , DeCBOR (TxCert era)
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/BlockChain.hs:
```diff
@@ -188,14 +191,16 @@ bbHash (TxSeq' _ bodies wits md) =
+constructMetadata = indexLookupSeq
+
+indexLookupSeq :: Int -> Map Int a -> Seq (Maybe a)
+indexLookupSeq n md = fmap (`Map.lookup` md) (Seq.fromList [0 .. n - 1])
```
It would have been also better to switch to `IntMap`, since it is a bit more efficient, but we can do it at some point later (created #4900)
```suggestion
indexLookupSeq n ixMap = Seq.fromList [Map.lookup ix ixMap | ix <- [0 .. n - 1]]
```

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxSeq/Internal.hs:
```diff
@@ -230,6 +233,65 @@ instance AlonzoEraTx era => DecCBOR (Annotator (AlonzoTxSeq era)) where
+    let mkTx body wit isValid aData =
+          mkBasicTx body
+            & witsTxL .~ wit
+            & auxDataTxL .~ aData
```
instead of mapping over a structure
```suggestion
            & auxDataTxL .~ maybeToStrictMaybe aData
```

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxSeq/Internal.hs:
```diff
@@ -230,6 +233,65 @@ instance AlonzoEraTx era => DecCBOR (Annotator (AlonzoTxSeq era)) where
+              ++ show (bodiesLength - 1)
+          )
+      )
+    let auxData = maybeToStrictMaybe <$> indexLookupSeq bodiesLength auxDataMap
```
Unlike lists `Seq` do not have fusion. Doing a `<$>` will construct a redundant `Seq` data structure
```suggestion
    let auxData = indexLookupSeq bodiesLength auxDataMap
```

teodanciu replied:
Interesting! Thanks

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Tx/Internal.hs:
```diff
@@ -348,33 +372,54 @@ unsafeConstructTxWithBytes b w a bytes = TxConstr (mkMemoBytes (ShelleyTxRaw b w
+  TxWits era ->
+  Maybe (TxAuxData era) ->
+  ShelleyTx era
+segWitTx body' witnessSet metadata =
+  let
+    wrappedMetadataBytes = case metadata of
+      Nothing -> Plain.serialize Plain.encodeNull
+      Just b -> Plain.serialize b
+    fullBytes =
+      Plain.serialize (Plain.encodeListLen 3)
+        <> Plain.serialize body'
+        <> Plain.serialize witnessSet
+        <> wrappedMetadataBytes
+   in
+    unsafeConstructTxWithBytes
+      body'
+      witnessSet
+      (maybeToStrictMaybe metadata)
+      fullBytes
```
I've now realized why this reserialization is here. It is pure for tx size calculation. I've created a ticket #4901 about it.

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/BlockChain.hs:
```diff
@@ -233,12 +238,45 @@ txSeqDecoder lax = do
+    Annotated bodies bodiesBytes <- decodeAnnotated decCBOR
+    Annotated wits witsBytes <- decodeAnnotated decCBOR
+    Annotated auxDataMap auxDataBytes <- decodeAnnotated decCBOR
+    let b = length bodies
```
Let's give it a proper name too
```suggestion
    let bodiesLength = length bodies
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/BlockChain.hs:
```diff
@@ -233,12 +238,45 @@ txSeqDecoder lax = do
+      (all inRange (Map.keysSet auxDataMap))
+      (fail ("Some Auxiliarydata index is not in the range: 0 .. " ++ show (b - 1)))
+    let auxData = indexLookupSeq b auxDataMap
+    let w = length wits
```
Ditto
```suggestion
    let witsLength = length wits
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/BlockChain.hs:
```diff
@@ -233,12 +238,45 @@ txSeqDecoder lax = do
+    let b = length bodies
+    let inRange x = (0 <= x) && (x <= (b - 1))
+    unless
+      (all inRange (Map.keysSet auxDataMap))
```
Same improvement can be applied here
```suggestion
      (getAll (Map.foldMapWithKey (\k _ -> All (inRange k)) auxDataMap))
```
In fact it feels like we could reduce come duplication by reusing some of this functionality in Alonzo with a bit of rearrangement. Feel free to tackle this bit of deduplication if you up to it.

teodanciu replied:
I made the suggested change for now and will follow with a PR to remove duplication  and the improvement you suggested for `indexLookupSeq`

lehins commented on libs/cardano-protocol-tpraos/src/Cardano/Protocol/TPraos/BHeader.hs:
```diff
@@ -292,6 +294,20 @@ instance Crypto c => DecCBOR (Annotator (BHeader c)) where
+instance Crypto c => DecCBOR (BHeader c) where
+  decCBOR = do
+    mb <- decodeMemoized decCBOR
+    let (Memo (BHeaderRaw bhb sig) bs) = mb
```
```suggestion
    Memo (BHeaderRaw bhb sig) bytes <- decodeMemoized decCBOR
```

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Arbitrary.hs:
```diff
@@ -888,6 +891,39 @@ instance Arbitrary Prices where
+instance
+  Era era =>
+  Arbitrary (Datum era)
+  where
```
Don't know why, but that vertical style formatting with one word per line always bugs me 
```suggestion
instance Era era => Arbitrary (Datum era) where
```

teodanciu replied:
I agree, it's horrible

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Keys/Bootstrap.hs:
```diff
@@ -125,13 +130,33 @@ instance EncCBOR BootstrapWitness
+
+instance DecCBOR BootstrapWitnessRaw where
+  decCBOR = decodeRecordNamed "BootstrapWitnessRaw" (const 4) $
+    do
```
Consistent style for `do` usage
```suggestion
  decCBOR = decodeRecordNamed "BootstrapWitnessRaw" (const 4) $ do
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Keys/Bootstrap.hs:
```diff
@@ -125,13 +130,33 @@ instance EncCBOR BootstrapWitness
+      pure . pure $ BootstrapWitness' key sig cc attributes
+
+data BootstrapWitnessRaw
+  = BootstrapWitnessRaw
+      !(VKey 'Witness)
+      !(SignedDSIGN DSIGN (Hash HASH EraIndependentTxBody))
+      !ChainCode
+      !ByteString
```
I think it already be useful to give them names
```suggestion
  = BootstrapWitnessRaw
  { bwrKey :: !(VKey 'Witness)
  , bwrSig :: !(SignedDSIGN DSIGN (Hash HASH EraIndependentTxBody))
  , bwrChainCode :: !ChainCode
  , bwrAttributes :: !ByteString
  }
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Keys/Bootstrap.hs:
```diff
@@ -125,13 +130,33 @@ instance EncCBOR BootstrapWitness
+      key <- decCBOR
+      sig <- decodeSignedDSIGN
+      cc <- decCBOR
+      BootstrapWitnessRaw key sig cc <$> decCBOR
```
Not sure if those names are that useful
```suggestion
      BootstrapWitnessRaw <$> decCBOR <*> decodeSignedDSIGN <*> decCBOR <$> decCBOR
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Keys/Bootstrap.hs:
```diff
@@ -125,13 +130,33 @@ instance EncCBOR BootstrapWitness
+instance DecCBOR BootstrapWitness where
+  decCBOR = do
+    mb <- decodeMemoized (decCBOR @BootstrapWitnessRaw)
+    let (Memo (BootstrapWitnessRaw k s c a) bs) = mb
```
```suggestion
    Memo (BootstrapWitnessRaw k s c a) bytes <- decodeMemoized (decCBOR @BootstrapWitnessRaw)
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Block.hs:
```diff
@@ -134,6 +136,37 @@ instance
+data BlockRaw h era = BlockRaw !h !(TxSeq era)
+
+instance
+  forall h era.
```
Unnecessary
```suggestion
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Block.hs:
```diff
@@ -134,6 +136,37 @@ instance
+      blockSize = 1 + fromIntegral (numSegComponents @era)
+
+instance
+  forall h era.
```
Ditto
```suggestion
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Block.hs:
```diff
@@ -134,6 +136,37 @@ instance
+  where
+  decCBOR = do
+    mb <- decodeMemoized (decCBOR @(BlockRaw h era))
+    let (Memo (BlockRaw h txSeq) bs) = mb
```
```suggestion
    Memo (BlockRaw h txSeq) bytes <- decodeMemoized (decCBOR @(BlockRaw h era))
```

lehins reviewed (approved):
This is awesome!
All the comments that I posted are just non-critical suggestions
Once you address them and figure out the test failure this PR will be ready to go.
